cookbooks/cdo-secrets/recipes/app.rb

```diff
@@ -0,0 +1,19 @@
+# Load deployment.rb into Chef context.
+# This makes CDO.* variables and secrets available to Chef resources in the execution phase.
+#
+# To resolve a reference in the execution phase use `#lazy in a Chef-resource property, e.g.:
+#
+# resource do
+#   property lazy {CDO.variable_name}
+# end
+
+# Install gem dependencies used by Cdo::Secrets.
+chef_gem 'aws-sdk-secretsmanager'
+chef_gem 'activesupport'
+
+ruby_block 'CDO config' do
+  block do
+    ENV['BUNDLE_GEMFILE'] = '' # Disable Bundler
+    require "#{node['cdo-repository']['git_path']}/deployment"
```

Comment:
Still trying to wrap my head around how this works. It seems like this is the crux of the change, right? This makes `CDO` available to all other recipes via being a global Ruby variable, and not through any Chef-specific mechanisms?

I take it this isn't the normal "Chef Way" of doing things, though I'm not sure exactly what that would be. Is the normal way to set this kind of data as attributes on the node? 

Is there something about our use case which is different from the normal use case? Or are we differing because this the easiest / cleanest way to implement it because there's so much going on with the `CDO` configuration object?

Reply:
Yes, this is the line of code that matters. It makes `CDO.*` variables accessible from Chef code

The normal 'Chef Way' of doing things would be to keep 'configuration' in node [attributes](https://docs.chef.io/nodes/#attributes) and 'secrets' in [encrypted data bags](https://docs.chef.io/secrets/), where both attributes and secrets are manually maintained in a hosted Chef repo and/or populated through cookbooks/recipes. In this model, Chef and its APIs are the primary source of configuration and secrets.

Our use case has always been different from this 'normal' use case because we maintain our Chef environment as a smaller component of our larger monolithic application (`/cookbooks`), not the other way around (where e.g., one Chef environment provisions and manages various smaller applications). Our use case has always been a much better fit for running Chef in local-mode, treating it as a simpler, narrower, self-contained provisioning library that consumes its configuration and secrets directly from its parent application. This PR is related to this longer-term migration process.

I had considered adding a 'configuration integration' cookbook (and/or Ohai plugin) that would import our configuration into Chef attributes that could then be accessed from cookbooks using the Chef-standard node-attribute APIs, but that would have been an extra chunk of non-trivial work, it would make it not possible to lazy-load specific secrets when referenced (it would have to fetch all secrets on every Chef-client run to make them available), and it would only make it more difficult to reference application config/secrets from Chef code (because the API would be different from how the rest of our application references the same config).

Reply:
Sounds good, thanks for the detailed explanation. I understand that we want to use chef local mode and why. This approach makes sense.

I think I disagree that the provisioning process is a component of our application. I see it more as our application and the provisioning process now both depending on the `deployment` component. (Maybe this is obvious - I think we've been following this pattern for other things like scripts also - but I think it's worth articulating.)

I think it'd be helpful if you could add comments `deployment.rb` to describe its general role and name some of the places it's depended on, including now our Chef cookbooks.

Reply:
> I think it'd be helpful if you could add comments `deployment.rb` to describe its general role and name some of the places it's depended on, including now our Chef cookbooks.

Agree in general, I have a few additional ideas on improvements around this piece specifically that I'll work on together with a comment/documentation pass.